servlet/src/main/java/com/linecorp/armeria/server/servlet/DefaultServletHttpRequest.java

```diff
@@ -0,0 +1,727 @@
+
+import static com.google.common.base.Strings.isNullOrEmpty;
+import static java.util.Objects.requireNonNull;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.io.UnsupportedEncodingException;
+import java.net.InetSocketAddress;
+import java.nio.charset.Charset;
+import java.security.Principal;
+import java.text.DateFormat;
+import java.text.SimpleDateFormat;
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Date;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+import javax.annotation.Nullable;
+import javax.servlet.AsyncContext;
+import javax.servlet.DispatcherType;
+import javax.servlet.ServletException;
+import javax.servlet.ServletRequest;
+import javax.servlet.ServletResponse;
+import javax.servlet.SessionTrackingMode;
+import javax.servlet.http.Cookie;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.HttpSession;
+import javax.servlet.http.HttpUpgradeHandler;
+import javax.servlet.http.Part;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+
+import com.linecorp.armeria.common.AggregatedHttpRequest;
+import com.linecorp.armeria.common.HttpHeaderNames;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.server.ServiceRequestContext;
+
+import io.netty.buffer.Unpooled;
+
+/**
+ * The servlet request.
+ */
+final class DefaultServletHttpRequest implements HttpServletRequest {
+    private static final Logger logger = LoggerFactory.getLogger(DefaultServletHttpRequest.class);
+    private static final Locale[] DEFAULT_LOCALS = { Locale.getDefault() };
+    private static final String RFC1123_DATE = "EEE, dd MMM yyyy HH:mm:ss zzz";
+    private static final SimpleDateFormat[] FORMATS_TEMPLATE = {
+            new SimpleDateFormat(RFC1123_DATE, Locale.ENGLISH),
+            new SimpleDateFormat("EEEEEE, dd-MMM-yy HH:mm:ss zzz", Locale.ENGLISH),
+            new SimpleDateFormat("EEE MMMM d HH:mm:ss yyyy", Locale.ENGLISH)
+    };
+
+    private final ServiceRequestContext serviceRequestContext;
+    private final DefaultServletContext servletContext;
+    private final AggregatedHttpRequest httpRequest;
+    private final DefaultServletInputStream inputStream;
+    private final Map<String, Object> attributeMap = new ConcurrentHashMap<>(16);
+    private final SessionTrackingMode sessionIdSource = SessionTrackingMode.COOKIE;
+
+    private final LinkedMultiValueMap<String, String> parameterMap = new LinkedMultiValueMap<>();
```

Comment:
Is there any chances that two thread access this map?

Reply:
I think two thread never access one this map instance because each httprequest run on one thread and create a parameterMap instance.